ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,160 @@
+    (fun msg ->
+      error "%s" msg ;
+      raise Api_errors.(Server_error (internal_error, [msg]))
+    )
+    fmt
+
+let drivertool args =
+  let path = !Xapi_globs.driver_tool in
+  try
+    let stdout, _stderr = Forkhelpers.execute_command_get_output path args in
+    debug "%s: executed %s %s" __FUNCTION__ path (String.concat " " args) ;
+    stdout
+  with e ->
+    internal_error "%s: failed to run %s %s: %s" __FUNCTION__ path
+      (String.concat " " args) (Printexc.to_string e)
+
+module Variant = struct
+  let create ~__context ~name ~version ~driver ~hw_present ~priority ~dev_status
+      =
+    info "%s: %s" __FUNCTION__ name ;
+    let ref = Ref.make () in
+    let uuid = Uuidx.to_string (Uuidx.make ()) in
+    Db.Driver_variant.create ~__context ~ref ~driver ~uuid ~name ~version
+      ~hardware_present:hw_present ~priority ~status:dev_status ;
+    ref
+
+  let destroy ~__context ~self =
+    debug "Destroying driver variant %s" (Ref.string_of self) ;
+    Db.Driver_variant.destroy ~__context ~self
+
+  let select ~__context ~self =
+    debug "%s: %s" __FUNCTION__ (Ref.string_of self) ;
+    let drv = Db.Driver_variant.get_driver ~__context ~self in
+    let d = Db.Host_driver.get_record ~__context ~self:drv in
+    let v = Db.Driver_variant.get_record ~__context ~self in
+    let stdout =
+      drivertool ["select"; d.API.host_driver_name; v.API.driver_variant_name]
+    in
+    info "%s: %s" __FUNCTION__ stdout ;
+    Db.Host_driver.set_selected_variant ~__context ~self:drv ~value:self
+end
+
+let create ~__context ~host ~name ~friendly_name ~_type ~description ~info:inf
+    ~active_variant ~selected_variant =
+  D.info "%s: %s" __FUNCTION__ name ;
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
+  Db.Host_driver.create ~__context ~ref ~uuid ~host ~name ~friendly_name ~_type
+    ~description ~info:inf ~active_variant ~selected_variant ;
+  ref
+
+(** destroy driver and recursively its variants, too *)
+let destroy ~__context ~self =
+  D.debug "Destroying driver %s" (Ref.string_of self) ;
+  let variants = Db.Host_driver.get_variants ~__context ~self in
+  variants |> List.iter (fun self -> Variant.destroy ~__context ~self) ;
+  Db.Host_driver.destroy ~__context ~self
+
+(** Runs on the host where the driver is installed *)
+let select ~__context ~self ~variant =
+  let drv = Ref.string_of self in
+  let var = Ref.string_of variant in
+  let variants = Db.Host_driver.get_variants ~__context ~self in
+  match List.mem variant variants with
+  | true ->
+      D.debug "%s selecting driver %s variant %s" __FUNCTION__ drv var ;
+      let d = Db.Host_driver.get_record ~__context ~self in
+      let v = Db.Driver_variant.get_record ~__context ~self:variant in
+      let stdout =
+        drivertool ["select"; d.API.host_driver_name; v.API.driver_variant_name]
+      in
+      info "%s: %s" __FUNCTION__ stdout ;
+      Db.Host_driver.set_selected_variant ~__context ~self ~value:variant
```

Comment:
Looks like a copy of `Variant.select`, except you have the driver ref already. But I'm not sure we actually need `Driver_variant.select` in the API at all.

Reply:
One uses the name of the variant, the other the ref. Both may be the most convenient for an API client; I would make the decision to remove this later as the feature progresses.

Reply:
So both for `select` and `rescan` have redundant versions in the datamodel now. I prefer to keep the API simple and have just one way of doing something, and keep just one of each kind. But fine if you want to decide later which one is best.